### PR TITLE
Split Source model to Source and UAST

### PR DESCRIPTION
**Pull request by zurk**

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
Coverage dropped

File: ast2vec/repo2/source.py [hunk 1]
```diff
@@ -24,9 +20,13 @@ def convert_uasts(self, file_uast_generator):
...
             sources = self._get_sources(file_uast.filepath)
-            src_codes.append(sources)
-            uasts.append(file_uast.response)
-            filenames.append(file_uast.filename)
+            if sources is not None:
+                src_codes.append(sources)
+                uasts.append(file_uast.response)
+                filenames.append(file_uast.filename)
+
+        assert len(src_codes) == len(uasts)
```
Comment (vmarkovtsev) [inline ast2vec/repo2/source.py]: Assertion is smth which is a bug in your code. Here you validate the file which comes from the outer world, so you should throw an exception instead. `ValueError` I guess

File: ast2vec/repo2/source.py [hunk 2]
```diff
@@ -24,9 +20,13 @@ def convert_uasts(self, file_uast_generator):
 
         for file_uast in file_uast_generator:
             sources = self._get_sources(file_uast.filepath)
```
Comment (vmarkovtsev) [inline ast2vec/repo2/source.py]: `_get_source`

File: ast2vec/repo2/source.py [hunk 3]
```diff
@@ -24,9 +20,13 @@ def convert_uasts(self, file_uast_generator):
...
-            src_codes.append(sources)
-            uasts.append(file_uast.response)
-            filenames.append(file_uast.filename)
+            if sources is not None:
```
Comment (vmarkovtsev) [inline ast2vec/repo2/source.py]: if source is None: continue

File: ast2vec/repo2/source.py [hunk 4]
```diff
@@ -46,8 +46,6 @@ def dependencies(self):
 
     def result_to_model_kwargs(self, result, url_or_path):
         filenames, src_codes, uasts = result
-        return {
-            "filenames": filenames,
-            "sources": src_codes,
-            "uasts": uasts
-        }
+        if len(filenames) == 0:
+            raise ValueError('No need to store empty model.')
```
Comment (vmarkovtsev) [inline ast2vec/repo2/source.py]: "The model is empty"

File: ast2vec/source.py [hunk 5]
```diff
@@ -1,27 +1,26 @@
-from bblfsh.github.com.bblfsh.sdk.protocol.generated_pb2 import ParseResponse
-from modelforge import generate_meta
-from modelforge.model import Model, split_strings, merge_strings, write_model
+from modelforge.model import split_strings, merge_strings
 from modelforge.models import register_model
 
-import ast2vec
+from ast2vec.uast import UAST
```
Comment (vmarkovtsev) [inline ast2vec/source.py]: We need to rename UAST to UASTModel, otherwise it is too confusing

File: ast2vec/source.py [hunk 6]
```diff
@@ -1,27 +1,26 @@
...
+class Source(UAST):
     """
     Model for source-code storage
     """
     NAME = "source"
 
     def construct(self, filenames, sources, uasts):
-        self._filenames = filenames
+        super(Source, self).construct(filenames=filenames, uasts=uasts)
+        assert len(sources) == len(uasts)
```
Comment (vmarkovtsev) [inline ast2vec/source.py]: ValueError

File: ast2vec/source.py [hunk 7]
```diff
@@ -36,26 +35,23 @@ def sources(self):
...
-        self._meta = generate_meta(self.NAME, ast2vec.__version__, *deps)
-        write_model(self._meta,
-                    {"filenames": merge_strings(self.filenames),
-                     "sources": merge_strings(self.sources),
-                     "uasts": [uast.SerializeToString() for uast in self.uasts]},
-                    output)
+        for item in zip(self._filenames, *super(Source, self).__iter__()):
+            yield item
+
+    def _to_save_dict(self):
```
Comment (vmarkovtsev) [inline ast2vec/source.py]: _to_dict_to_save

File: ast2vec/uast.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+from modelforge.model import Model, split_strings, merge_strings, write_model
+from modelforge.models import register_model
+
+import ast2vec
+
+
+@register_model
+class UAST(Model):
+    """
+    Model for uast storage
```
Comment (vmarkovtsev) [inline ast2vec/uast.py]: to store Univeral Abstract Syntax Trees

File: ast2vec/uast.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        :param item: File index.
+        :return: name, uast
+        """
+        return self._filenames[item], self._uasts[item]
+
+    def __iter__(self):
+        """
+        Iterator over the items.
+        """
+        for item in zip(self._filenames, self._uasts):
```
Comment (vmarkovtsev) [inline ast2vec/uast.py]: return iter(zip(...))

File: ast2vec/source.py [hunk 10]
```diff
@@ -36,26 +35,23 @@ def sources(self):
...
-    def save(self, output, deps=None):
-        if not deps:
-            deps = tuple()
-        self._meta = generate_meta(self.NAME, ast2vec.__version__, *deps)
-        write_model(self._meta,
-                    {"filenames": merge_strings(self.filenames),
-                     "sources": merge_strings(self.sources),
-                     "uasts": [uast.SerializeToString() for uast in self.uasts]},
-                    output)
+        for item in zip(self._filenames, *super(Source, self).__iter__()):
```
Comment (vmarkovtsev) [inline ast2vec/source.py]: return iter(zip())

File: ast2vec/uast.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        return {"filenames": merge_strings(self.filenames),
+                "uasts": [uast.SerializeToString() for uast in self.uasts]}
+
+    def save(self, output, deps=None):
+        if not deps:
+            deps = tuple()
+        self._meta = generate_meta(self.NAME, ast2vec.__version__, *deps)
+        write_model(self._meta, self._to_save_dict(), output)
+
+    def become_uast(self, vocabulary):
```
Comment (vmarkovtsev) [inline ast2vec/uast.py]: Remove this - not relevant

File: ast2vec/uast.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+@register_model
+class UASTModel(Model):
+    """
+    Model to store Univeral Abstract Syntax Trees
+    """
+    NAME = "uast"
+
+    def construct(self, filenames, uasts):
+        if not len(uasts) == len(filenames):
```
Comment (fineguy) [inline ast2vec/uast.py]: `if len(uasts) != len(filenames):`

File: ast2vec/uast.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            raise ValueError("Length of uasts({}) and filenames({}) is not equal".
```
Comment (fineguy) [inline ast2vec/uast.py]: `raise ValueError("Lengths of uasts({}) and filenames({}) are not equal".`

File: ast2vec/source.py [hunk 14]
```diff
@@ -1,27 +1,28 @@
...
     """
     Model for source-code storage
     """
     NAME = "source"
 
     def construct(self, filenames, sources, uasts):
-        self._filenames = filenames
+        super(Source, self).construct(filenames=filenames, uasts=uasts)
+        if not len(sources) == len(uasts) == len(filenames):
+            raise ValueError("Length of src_codes({}), uasts({}) and filenames({}) is not equal".
```
Comment (fineguy) [inline ast2vec/source.py]: `raise ValueError("Lengths of src_codes({}), uasts({}) and filenames({}) are not equal".`

File: ast2vec/uast.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    def dump(self):
+        symbols_num = 100
+        out = self._uasts[0][:symbols_num]
+        return "Number of files: %d. First %d symbols:\n %s" % (
+            len(self), symbols_num, out)
+
+    @property
+    def uasts(self):
+        """
+        Returns all usts of code in the saved repo
```
Comment (fineguy) [inline ast2vec/uast.py]: `Return all uasts of code in the saved repo`

**Pull request merged**